app/Models/Foundation/Summit/Registration/PromoCodes/SummitRegistrationPromoCode.php

```diff
@@ -667,4 +701,74 @@ public function clearTags()
     {
         $this->tags->clear();
     }
+
+    /**
+     * @return SummitAttendeeTicket[]
+     */
+    public function getTickets()
+    {
+        return $this->tickets;
+    }
+
+    /**
+     * @return SummitAttendeeTicket[]
+     */
+    public function getUnassignedTickets()
+    {
+        $criteria = Criteria::create();
+        $criteria->where(Criteria::expr()->isNull('owner'));
+        return $this->tickets->matching($criteria);
+    }
+
+    /**
+     * @param SummitAttendeeTicket $ticket
+     */
+    public function addTicket(SummitAttendeeTicket $ticket): void
+    {
+        if ($this->tickets->contains($ticket)) return;
+        $ticket->setPromoCode($this);
+        $this->tickets->add($ticket);
+    }
+
+    /**
+     * @param SummitTicketType $ticket_type
+     * @return SummitAttendeeTicket|null
+     */
+    public function getNextAvailableTicketPerType(SummitTicketType $ticket_type): ?SummitAttendeeTicket
+    {
+        $excluded_ids = [];
+
+        do {
+            $query = <<<SQL
+SELECT e  
+FROM models\summit\SummitAttendeeTicket e
+WHERE 
+e.ticket_type = :type
+AND e.owner IS NULL
+AND e.id NOT IN (:excluded_ids)
+SQL;
+
+            $native_query = $this->getEM()->createQuery($query);
+            $native_query->setParameter("type", $ticket_type);
+            $native_query->setParameter("excluded_ids", implode(',', $excluded_ids));
+            $ticket =  $native_query->getOneOrNullResult();
+
+            if(is_null($ticket)) break;
+            $key = sprintf("prepaid.ticket.%s.lock", $ticket->getId());
+            if (Cache::has($key)) {
+                $excluded_ids[] = $ticket->getId();
+                continue;
+            }
+            // lock
+            Cache::put($key, $key, now()->addMinutes(5));
```

Comment:
move 5 to constant def on the class like LOCK_SKEW_TIME or something else

Reply:
done